design-proposals/feature-lifecycle.md

```diff
@@ -0,0 +1,327 @@
+  for the project contributors.
+  > **Note**: As of this writing (pre v1.2), there are 37 FGs,
+  > out of which 5 GA-ed and 2 marked for deprecation.
+  > For more information, explore the
+  > [source](https://github.com/kubevirt/kubevirt/blob/5ff12ae931cefd81514ec96f97a189ab2c179ad7/pkg/virt-config/feature-gates.go).
+- We do not have agreed-upon procedures on how and when features
+  graduate or discontinue.
+  This causes each feature to take different approaches, possibly
+  surprising users.
+
+We conclude that the current use of FGs is insufficient
+due to the lack of well-defined processes and policy on how a feature
+should progress in its lifecycle.
+
+Eventually we would like to see features being evaluated carefully
+before they are introduced, while they are experimented with and
+proven to be actually in use (and useful) before graduating.
+
+> **Note**: Once a feature graduates, it is included in a
+> General Availability (GA) release with its functionality available
+> to all users. GA features need to comply with [semver](https://semver.org/)
+> which add constraints on their ability to change (including deprecation).
+
+### Goals
+- Define the process a feature needs to pass in order to be
+  Generally Available.
+- Define the process a feature needs to pass in order to be removed.
+- Provide policies and rules on how to manage a feature during its
+  lifetime.
+
+### Non Goals
+- Implement enforcement tooling to keep features in sync with
+  the lifecycle rules.
+
+### Definition Of Users
+- Development contributors.
+- Cluster operators.
+
+### User Stories
+- As a KubeVirt contributor, I would like to introduce a new useful
+  feature and follow it to graduation (GA).
+- As a KubeVirt contributor, I would like to remove a feature
+  that has not yet reached its formal graduation.
+- As a KubeVirt contributor, I would like to remove a feature
+  that has already graduated.
+  > **Note**: Removal of a GA feature is considered an exception.
+  > Strong arguments and a wide agreement is required for such an action.
+- As a KubeVirt cluster operator, I would like to experiment with
+  a newly-proposed ("Alpha") feature in a controlled environment, to see
+  if it makes sense to me.
+- As a KubeVirt cluster operator, I would like to evaluate an
+  undergraduate ("Beta") feature in a real-life environment with actual users.
+- As a KubeVirt cluster operator, I would like to keep using a feature
+  that got graduated after I used it during the "Beta" evaluation period.
+- As a KubeVirt cluster operator, I would like to know that
+  an experimental ("Alpha") feature is planned to be removed.
+- As a KubeVirt cluster operator, I would like to know that
+  an undergraduate ("Beta") feature is planned to be removed.
+- As a KubeVirt cluster operator, I would like to stop using a feature
+  that got removed.
+
+### Repos
+This is a cross repo project policy under the
+[kubevirt](https://github.com/kubevirt) organization.
+
```

Comment:
I suggest we refine the proposal to apply it first within kubevirt/kubevirt before extending it to other repositories. Some projects are still early in development, with limited usage and contributors.

Reply:
The focus is on v1 APIs

> The proposal is focusing on introducing features in
> a stable API (CRD) version, e.g. `kubevirt.io/v1`.

As an organization I think everyone should be aware of the expected process for a graduated API.

This proposal is attempting to give a direction, but it does not strongly enforces it on a specific project.
The wording has been intentionally made in a manner that leaves room for movement.

E.g.
> A feature is expected to pass in the following order through the following stages:

If others will push in this direction, to reduce the scope, I will adjust.
@fabiand , any thoughts about this from your side?

Reply:
I'm a little confused because you said we treat the word "feature" like a black box, but it seems that now, it actually has a stable API version inside.

Reply:
I believe we should start by applying this new approach in 'kubevirt/kubevirt' first. If it proves useful (which I am optimistic about), other repositories may also choose to adopt it.